bentoml/metrics.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+from abc import ABC
+from abc import abstractmethod
+from typing import TYPE_CHECKING
+
+from simple_di import inject
+from simple_di import Provide
+
+logger = logging.getLogger(__name__)
+
+from ._internal.configuration.containers import BentoMLContainer
+
+if TYPE_CHECKING:
+    from bentoml._internal.server.metrics.prometheus import PrometheusClient
+
+
+class BaseMetric(ABC):
+    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
+        if "registry" in kwargs:
+            logger.warning(
+                "'registry' shouldn't be passed as an argument as BentoML will handle multiprocess mode for prometheus metrics. Removing for now..."
+            )
+            kwargs.pop("registry")
+        self._args = args
+        self._kwargs = kwargs
+        self._metric = None
+
+    @abstractmethod
+    def _setup_metric(self, metric_client: PrometheusClient) -> None:
+        raise NotImplementedError
+
+    def __getattr__(self, item: t.Any):
+        if self._metric is None:
+            self._setup_metric()  # type: ignore
```

Comment:
What is the reason for `# type: ignore` here?

Reply:
I get warning from type checker because `metric_client` is not marked as optional.

Reply:
the type warning is correct, which means the ignore is not correct.